@navikt/core/react/src/modal/Modal.tsx

```diff
@@ -79,111 +89,134 @@ type ModalLifecycle = {
 
-    const onModalCloseRequest = (e) => {
-      if (shouldCloseOnOverlayClick || e.type === "keydown") {
-        onClose();
-      } else if (buttonRef.current) {
-        buttonRef.current.focus();
+    useEffect(() => {
+      if (needPolyfill && modalRef.current) {
+        dialogPolyfill.registerDialog(modalRef.current);
       }
-    };
+    }, [modalRef]);
 
-    const getParentSelector = () => {
-      if (parentSelector) {
-        return parentSelector;
+    useEffect(() => {
+      // We need to have this in a useEffect so that the content renders before the modal is displayed,
+      // and in case `open` is true initially.
+      if (modalRef.current && open !== undefined) {
+        if (open && !modalRef.current.open) {
+          modalRef.current.showModal();
+        } else if (!open && modalRef.current.open) {
+          modalRef.current.close();
+        }
       }
-      return rootElement !== undefined
-        ? () => rootElement as HTMLElement
-        : undefined;
-    };
+    }, [modalRef, open]);
 
     return (
-      <ReactModal
-        {...rest}
-        parentSelector={getParentSelector()}
-        style={style}
-        isOpen={open}
+      <dialog
         ref={mergedRef}
-        className={cl("navds-modal", className)}
-        overlayClassName={cl("navds-modal__overlay", overlayClassName)}
-        shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
-        shouldCloseOnEsc={shouldCloseOnEsc}
-        onRequestClose={(e) => onModalCloseRequest(e)}
-        aria={{
-          describedby: ariaDescribedBy,
-          labelledby: ariaLabelledBy,
-          modal: ariaModal,
+        className={cl("navds-modal", className, {
+          "navds-modal--polyfilled": needPolyfill,
+        })}
+        onCancel={(event) => {
+          // FYI: onCancel fires when you press Esc
+          if (onBeforeClose && onBeforeClose() === false) {
+            event.preventDefault();
+          } else if (onCancel) onCancel(event);
         }}
-        contentLabel={contentLabel}
+        aria-labelledby={
+          ariaLabelledBy ?? (header ? internalAriaId : undefined)
+        }
+        {...rest}
       >
-        {children}
-        {closeButton && (
-          <Button
-            className={cl("navds-modal__button", {
-              "navds-modal__button--shake": shouldCloseOnOverlayClick,
-            })}
-            size="small"
-            variant="tertiary-neutral"
-            ref={buttonRef}
-            onClick={onClose}
-            icon={<XMarkIcon title="Lukk modalvindu" />}
-          />
-        )}
-      </ReactModal>
+        <ModalContext.Provider
+          value={{
+            closeHandler: getCloseHandler(modalRef, header, onBeforeClose),
+          }}
+        >
+          {header && (
+            <ModalHeader>
+              {header.label && (
+                <Detail
+                  as="label"
+                  className="navds-modal__label"
+                  id={internalAriaId}
```

Comment:
Tror hvis label er lagt til at den heller bør ha sin egen `id` + settes som `aria-describedby` på dialog (men må testes for hva som er best uu)

Reply:
Litt usikker på hva du mener her. Mener du at `heading` skal settes som `aria-labelledby` og `label` som `aria-describedby`? På mdn står det at "a label should be concise, while a description is intended to provide more verbose information".

Reply:
Ja tenker slik du skrev det, selve headingen er blir da tittelen på modalen, mens label gjerne bare er ekstra metadata/kategorisering

Reply:
Jeg tenker at aria-labelledby bør være ganske kort, da den (iallfall på min maskin) blir lest opp to ganger. Det er jo vanskelig å forutsi hva folk kommer til å gjøre i praksis, men jeg tror header.label er tenkt å være ganske kort (som en ekstra kategorisering/metadata som du skriver). Derfor tenker jeg at denne kanskje kan passe bedre som aria-labelledby, og så kan heading være aria-describedby (men labelledby hvis heading.label er tom).

Se på dette eksempelet fra virkeligheten: https://nav-it.slack.com/archives/C7NE7A8UF/p1683635163568579?thread_ts=1683634432.272119&cid=C7NE7A8UF
Hvis jeg tolker innholdet riktig, så er "Tiltak gjennom NAV..." mer en tittel på meldingen som skal sendes, enn på selve modalen. Nå er jo dette bare ett eksempel, så kan jo hende dette er mer et unntak enn regelen.

Jeg blir egentlig mer usikker jo mer jeg tenker på dette 😅. Mye mulig min logikk er "flawed" her, så det er helt greit om du fortsatt synes labelen skal være aria-describedby. En fordel med det er jo enklere logikk, som er enklere å forklare/forstå for brukerne (utviklerne). Kan forresten også høre med designerne i morgen hvordan de tenker labelen skal brukes.

Reply:
👌 Kan sikkert ta en prat med @it-vegard å høre hva som vil fungere best her, ikke helt sikker selv

Reply:
Linda og Jørgen mente at labelen var mer sekundær (og kommer trolig ikke til å bli brukt så ofte), så da heller jeg mot ditt forslag 😊